Add definition of specification development stages (#543)
slsa-framework/slsa: docs/spec-stages.md

---
title: Specification Stages

> layout: specifications

Reviewer: Should this document be part of the spec folder? Should it be versioned along with the rest of the spec (so that we can change it in newer spec versions, if we need to)?

Author: I have to admit that I wasn't sure about that but I think it should apply to all specifications, including the Provenance spec for instance. So it makes sense to have it in a central place.
And as far as versioning is concerned I don't know that we would need to have the different versions available concurrently so until the need arise I wouldn't bother.

Reviewer 2: I think it makes sense where it is, since it applies to all versions of all specifications.

---

> Specifications go through various stages of development from which you
> should have different expectations. This document defines these
> different stages and their meaning.

Reviewer: ```suggestion
Specifications go through various stages of development from which you
should have different expectations. This document defines the different
stages the SLSA project uses and their meaning for readers and
contributors.
```

> Every specification should prominently display a *Status of this
> document* section stating which stage the specification you are
> looking at is in.

Reviewer: I think SLSA is a specification, not multiple specifications. Let's refer to individual pages (documents? sections?) of the spec and not individual specs.
```suggestion
Every page of the specification should prominently display a *Status of this
document* section stating which stage the specification page is in.
```

Author: I was thinking of other specifications like Provenance. Isn't that a specification too?
Besides, I think we're better off not getting too much into the details of how this gets implemented so we keep more freedom on that front: whether to add a link, a popup, an actual section, etc.

Reviewer: That's a good point, we do imply that provenance and vsa are separate specifications.

Reviewer 2: I prefer Joshua's suggestion - every page should display the status. But I suggest that it's just "Status" rather than "Status of this document" since every page within the spec shares the same status (Arnuad's interpretation, below). Maybe also append "linking to the relevant section here"?

Author: I'm happy to change to just "Status" and I agree with having it on every page with a link to the definition. I wouldn't get that level of details on how we implement this here though but I will add it since you guys prefer it.

## Working Draft

> This is the first stage of development a specification goes
> through. At this point, not much should be expected of it. The
> specification may be very incomplete and may change at any time and
> even be abandoned. It is therefore not suitable for reference or for
> implementation beyond experimentation.

Reviewer: specification --> specification page
```suggestion
This is the first stage of development a page of the specification goes
through. At this point, not much should be expected of it. The page may
be very incomplete and may change at any time and even be abandoned. It is
therefore not suitable for reference or for implementation beyond
experimentation.
```

Author: ok, so we seem to have a different mental model here. You're thinking of allowing each page to have its own stage?
I was think of the specification as a whole.

Reviewer 2: I agree with @lehors and prefer the original wording.

> A specification may be published several times during this stage as
> work progresses. The status section of the document may provide
> additional information as to its development status and whether
> reviews and feedback are welcome.

Reviewer: ```suggestion
A specification page may be published several times during this stage as
work progresses. The status section of the document may provide
additional information as to its development status and whether
reviews and feedback are welcome.
```

Author: same here (mental model/granularity)

Reviewer 2: same (agree with @lehors)

Reviewer: WFM

---
title: Specification Stages
layout: specifications
---

Specifications go through various stages of development from which you
should have different expectations. This document defines the different
stages the SLSA project uses and their meaning for readers and
contributors.

Every specification should prominently display a *Status of this
document* section stating which stage the specification you are
looking at is in.

## Working Draft

This is the first stage of development a specification goes
through. At this point, not much should be expected of it. The
specification may be very incomplete and may change at any time and
even be abandoned. It is therefore not suitable for reference or for
implementation beyond experimentation.

A specification may be published several times during this stage as
work progresses. The status section of the document may provide
additional information as to its development status and whether
reviews and feedback are welcome.

> ## Proposed Specification

Reviewer 2: nit: I find "Specification" in the status to be awkward. (I realize that this matches the W3C and IETF usage, but it doesn't seem to fit well here.)

Any thoughts on any of the following alternatives?

- "Proposed" or "In Review"
- "Published", "Accepted", or "Stable"

Author: For what it's worth that's not how W3C and IETF call them but we're not bound to them anyway so no problem there. Plus I have to admit I wasn't sure about using "Specification" in the status either so I'm good with dropping it.
I don't really feel strongly about those but I quite like "In Review" and "Stable".

Reviewer 2: 👍

At this stage the document is considered to be feature complete and is
published as a way to invite final reviews. Editorial changes may
still be made but no addition of new features is expected and short of
problems being found no significant changes are expected to happen
anymore.

## Specification

At this stage the document is considered final and stable. It is
suitable for reference and implementation in production. Changes may
only happen if deemed necessary and will be avoided as much as

> possible if they might adversely impact implementations.

Reviewer 2: I think "final" and the restrictions on changes are a bit too restrictive. We may continue to clarify the specification without version bumps, but we won't change the meaning.

Author: Ok, I dropped final.

> ## Retired

Reviewer 2: Should we split "Outdated" from "Abandoned"? The former is still a perfectly valid specification that people can continue to rely on (just not the latest version) while the latter should not be used.

Author: It really depends on whether we are actually fine with people keeping on using an old version or whether we'd rather try to get them moving to the newer version. I was assuming the latter but it's worth some discussion.

Reviewer 2: From the meeting today:

- Keeping "Retired" is OK with me. Others may have an opinion, but I defer to Arnaud to make the decision.
- Let's soften the wording in the paragraph. Instead of "should not be used", rather something like "is no longer maintained" and something about new uses should prefer the newer version?

Author: Ok, I applied the change to "is no longer maintained". I think that should do.

> This stage indicates that the specification should no longer be
> used. It has been either rendered obsolete by a newer version or

Reviewer 3: Consider softening the language here:

```suggestion
This stage indicates that the specification is no longer maintained.
It has been either rendered obsolete by a newer version or
```
